### PR TITLE
Add support for multiple assembly dialects in the base parser

### PR DESCRIPTION
**Pull request by pleroy**
The parser and related code currently know about the ISA.  But of course the point of this effort is to support a different assembly dialect for the `x86` ISA.  While it would have been possible to lump the two pieces of information in a "fake ISA" (say, `x86ATT` and `x86INTEL`, this seems like an abuse of the data model.  A clean separation seems preferable.

This PR is adding the dialect (1) as a  command line flag and (2) in the heuristic detection done by the base parser.  It also plumbs the parameter everywhere necessary.  There is a new parser, but it doesn't implement any method other than a constructor at this point.

### CODE REVIEW

File: osaca/osaca.py [hunk 1]
```diff
@@ -47,6 +47,10 @@
     "aarch64": "V2",
     "x86": "SPR",
 }
+SUPPORTED_DIALECTS = {
```
Comment (eggrobin) [inline osaca/osaca.py]: Let’s call them syntaxes.

File: osaca/osaca.py [hunk 2]
```diff
@@ -310,38 +329,52 @@ def inspect(args, output_file=sys.stdout):
     code = args.file.read()
 
     # Detect ISA if necessary
-    arch = args.arch if args.arch is not None else DEFAULT_ARCHS[BaseParser.detect_ISA(code)]
+    detected_isa, detected_dialect = BaseParser.detect_ISA(code)
+    arch = args.arch if args.arch is not None else DEFAULT_ARCHS[detected_isa]
```
Comment (eggrobin) [inline osaca/osaca.py]:
```suggestion
    arch = args.arch if args.arch else DEFAULT_ARCHS[detected_isa]
```

File: osaca/osaca.py [hunk 3]
```diff
@@ -310,38 +329,52 @@ def inspect(args, output_file=sys.stdout):
...
+    dialect = args.dialect if args.dialect is not None else detected_dialect
```
Comment (eggrobin) [inline osaca/osaca.py]:
```suggestion
    dialect = args.dialect if args.dialect else detected_dialect
```

File: osaca/osaca.py [hunk 4]
```diff
@@ -310,38 +329,52 @@ def inspect(args, output_file=sys.stdout):
...
-            )
-            isa = MachineModel.get_isa_for_arch(arch)
-            parser = get_asm_parser(arch)
+    # Determine the combinations that should be tried based on the command line
+    # arguments.
+    combinations_to_try = ({
+        (DEFAULT_ARCHS["x86"], "ATT"),
+        (DEFAULT_ARCHS["x86"], "INTEL"),
+        (DEFAULT_ARCHS["aarch64"], None),
+    } if args.arch == None and args.dialect == None
```
Comment (eggrobin) [inline osaca/osaca.py]:
```suggestion
    } if not args.arch and not args.dialect
```

Same elsewhere.

**Review by eggrobin — CHANGES_REQUESTED** [review]

File: osaca/osaca.py [hunk 5]
```diff
@@ -232,6 +242,15 @@ def check_arguments(args, parser):
         parser.error(
             "Microarchitecture not supported. Please see --help for all valid architecture codes."
         )
+    if (args.syntax is not None and args.arch is not None and
```
Comment (eggrobin) [inline osaca/osaca.py]: More is not None.

**Review by eggrobin — APPROVED** [review]

File: osaca/osaca.py (hunk 5 above)
Comment (pleroy) [reply osaca/osaca.py]: Done, but note that this is the style above.  (I don't think that I should go cleaning up that code.)

**Review by eggrobin — APPROVED** [review]

**Pull request merged**